File: src/types/lca.ts
export interface LCAData {
  // Core Metal Data
  metalType: 'Aluminium' | 'Steel' | 'Copper' | 'Zinc' | 'Lead';
  
  // Extraction Tab
  oreGrade: number; // %
  miningMethod: 'Open Pit' | 'Underground' | 'Heap Leach';
  waterUsage: number; // m³/ton
  
  // Energy Tab
  totalEnergyConsumption: number; // kWh
  gridMix: {
    coal: number;
    hydro: number;
    solar: number;
    naturalGas: number;
  };
  processHeat: number; // MJ
  
  // Smelting Tab
  furnaceType: 'Electric Arc' | 'Blast' | 'Induction';
  temperature: number; // °C
  fluxUsage: number; // kg/ton
  slagRecovery: number; // %
  
  // Logistics Tab
  transportMode: 'Road' | 'Rail' | 'Sea' | 'Multi';
  inboundDistance: number; // km
  outboundDistance: number; // km
  vehicleEfficiency: number; // kg CO2/km
  
  // Circularity Tab
  scrapInputRate: number; // % (Recycled Content)
  recyclingEfficiency: number; // % (Scrap Recovery Rate)
  wasteRecovery: number; // %
  closedLoopRate: number; // %
  
  // Derived Values
  co2Emission: number; // kg
  circularityScore: number; // 0-100
  
  // Scenario
  scenarioType: 'Current' | 'Optimized' | 'Baseline';
}

export interface ScanResult {
  scrapType: string;
  purity: number;
  estimatedWeight: number;
  co2Saved: number;
  revenueEstimate: number;
  recommendedProcess: string;
}

export interface DoctorAnalysis {
  overallScore: number;
  carbonIntensity: {
    value: number;
    benchmark: number;
    isGap: boolean;
  };
  energyEfficiency: {
    value: number;
    benchmark: number;
    isGap: boolean;
  };
  circularityRating: string;
  improvements: DoctorImprovement[];
  riskFactors: string[];
}

export interface DoctorImprovement {
  id: string;
  title: string;
  description: string;
  potentialGain: number; // % improvement
  category: 'energy' | 'transport' | 'process' | 'circularity';
  simulateAction: Partial<LCAData>;
}

export interface MaterialPassport {
  id: string;
  generatedAt: string;
  lcaData: LCAData;
  doctorAnalysis: DoctorAnalysis;
  provenance: ProvenanceEvent[];
  certifications: string[];
  qrCode: string;
}

export interface ProvenanceEvent {
  timestamp: string;
  event: string;
  location: string;
  verified: boolean;
}

File: backend/app/models/lca.py
from pydantic import BaseModel, Field
from typing import Optional, Literal
from datetime import datetime
from bson import ObjectId

class PyObjectId(str):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate
    
    @classmethod
    def validate(cls, v, info):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid ObjectId")
        return str(v)

class GridMix(BaseModel):
    coal: float = Field(ge=0, le=100)
    hydro: float = Field(ge=0, le=100)
    solar: float = Field(ge=0, le=100)
    naturalGas: float = Field(ge=0, le=100)

class LCADataCreate(BaseModel):
    metalType: Literal['Aluminium', 'Steel', 'Copper', 'Zinc', 'Lead']
    oreGrade: float = Field(ge=10, le=100)
    miningMethod: Literal['Open Pit', 'Underground', 'Heap Leach']
    waterUsage: float = Field(ge=1, le=50)
    totalEnergyConsumption: float = Field(ge=1000, le=50000)
    gridMix: GridMix
    processHeat: float = Field(ge=1000, le=20000)
    furnaceType: Literal['Electric Arc', 'Blast', 'Induction']
    temperature: float = Field(ge=500, le=2000)
    fluxUsage: float = Field(ge=10, le=100)
    slagRecovery: float = Field(ge=0, le=100)
    transportMode: Literal['Road', 'Rail', 'Sea', 'Multi']
    inboundDistance: float = Field(ge=10, le=1000)
    outboundDistance: float = Field(ge=10, le=1000)
    vehicleEfficiency: float = Field(ge=0.01, le=1)
    scrapInputRate: float = Field(ge=0, le=100)
    recyclingEfficiency: float = Field(ge=50, le=100)
    wasteRecovery: float = Field(ge=0, le=100)
    closedLoopRate: float = Field(ge=0, le=100)
    scenarioType: Literal['Current', 'Optimized', 'Baseline'] = 'Current'

class LCADataResponse(LCADataCreate):
    id: str = Field(alias="_id")
    co2Emission: float
    circularityScore: float
    createdAt: datetime
    updatedAt: datetime
    
    class Config:
        populate_by_name = True

class LCADataUpdate(BaseModel):
    metalType: Optional[Literal['Aluminium', 'Steel', 'Copper', 'Zinc', 'Lead']] = None
    oreGrade: Optional[float] = None
    miningMethod: Optional[Literal['Open Pit', 'Underground', 'Heap Leach']] = None
    waterUsage: Optional[float] = None
    totalEnergyConsumption: Optional[float] = None
    gridMix: Optional[GridMix] = None
    processHeat: Optional[float] = None
    furnaceType: Optional[Literal['Electric Arc', 'Blast', 'Induction']] = None
    temperature: Optional[float] = None
    fluxUsage: Optional[float] = None
    slagRecovery: Optional[float] = None
    transportMode: Optional[Literal['Road', 'Rail', 'Sea', 'Multi']] = None
    inboundDistance: Optional[float] = None
    outboundDistance: Optional[float] = None
    vehicleEfficiency: Optional[float] = None
    scrapInputRate: Optional[float] = None
    recyclingEfficiency: Optional[float] = None
    wasteRecovery: Optional[float] = None
    closedLoopRate: Optional[float] = None
    scenarioType: Optional[Literal['Current', 'Optimized', 'Baseline']] = None
